src/Simplex/Chat.hs

```diff
@@ -1675,13 +1736,34 @@ processAgentMessage (Just user@User {userId, profile}) agentConnId agentMessage
                   toView $ CRReceivedContactRequest cReq
                   showToast (localDisplayName <> "> ") "wants to connect to you"
 
-    withAckMessage :: ConnId -> MsgMeta -> m () -> m ()
-    withAckMessage cId MsgMeta {recipient = (msgId, _)} action =
-      action `E.finally` withAgent (\a -> ackMessage a cId msgId `catchError` \_ -> pure ())
-
-    ackMsgDeliveryEvent :: Connection -> MsgMeta -> m ()
-    ackMsgDeliveryEvent Connection {connId} MsgMeta {recipient = (msgId, _)} =
-      withStore $ \db -> createRcvMsgDeliveryEvent db connId msgId MDSRcvAcknowledged
+    withCompletedCommand :: Connection -> ACommand 'Agent -> (CommandData -> m ()) -> m ()
+    withCompletedCommand Connection {connId} agentMsg action = do
+      cmdData_ <- withStore' $ \db -> getCommandDataByCorrId db user corrId
+      forM_ cmdData_ $ \cmdData@CommandData {cmdId, cmdConnId, cmdFunction} -> do
+        case cmdConnId of
+          Just cmdConnId' ->
+            if connId == cmdConnId'
+              then
+                if aCommandTag agentMsg == commandExpectedResponse cmdFunction
+                  then do
+                    withStore' $ \db -> updateCommandStatus db user cmdId CSCompleted
+                    action cmdData
+                  else throwChatError $ CECommandError "unexpected response for correlated command"
+              else throwChatError $ CECommandError "correlated command's connection id doesn't match"
+          Nothing -> throwChatError $ CECommandError "correlated command doesn't have connection id"
```

Comment:
not sure about re-using this error

Reply:
I would just combine all 4 conditions in one and throw one error otherwise.